File: apps/worker/src/workflows/promotion.workflow.ts
import * as wf from '@temporalio/workflow';
import type * as activities from '../activities/index.js';
import {
  abortSignal,
  gateFailedSignal,
  gatePassedSignal,
  pauseSignal,
  resumeSignal,
  statusQuery,
} from './signals.js';

const { persistRunState, recordAuditEvent, evaluateGate } = wf.proxyActivities<
  typeof activities
>({
  startToCloseTimeout: '30 seconds',
  retry: { maximumAttempts: 3 },
});

export interface PromotionWorkflowInput {
  promotionRunId: string;
  stageCount: number;
  actor: {
    actorType: 'user' | 'system' | 'api_key';
    actorId: string;
    displayName?: string;
  };
}

export async function promotionWorkflow(
  input: PromotionWorkflowInput,
): Promise<void> {
  let currentStageIndex = 0;
  let status: 'active' | 'paused' | 'aborted' | 'completed' = 'active';
  let isPaused = false;
  let gateAwaiting = false;

  wf.setHandler(statusQuery, () => ({ status, currentStageIndex, isPaused }));

  wf.setHandler(pauseSignal, async () => {
    isPaused = true;
    status = 'paused';
    await persistRunState({
      promotionRunId: input.promotionRunId,
      status: 'paused',
    });
    await recordAuditEvent({
      promotionRunId: input.promotionRunId,
      action: 'run_paused',
      actorType: 'system',
      actorId: 'workflow',
    });
  });

  wf.setHandler(resumeSignal, async () => {
    isPaused = false;
    status = 'active';
    await persistRunState({
      promotionRunId: input.promotionRunId,
      status: 'active',
    });
    await recordAuditEvent({
      promotionRunId: input.promotionRunId,
      action: 'run_resumed',
      actorType: 'system',
      actorId: 'workflow',
    });
  });

  wf.setHandler(abortSignal, async () => {
    status = 'aborted';
    await persistRunState({
      promotionRunId: input.promotionRunId,
      status: 'aborted',
    });
    await recordAuditEvent({
      promotionRunId: input.promotionRunId,
      action: 'run_aborted',
      actorType: 'system',
      actorId: 'workflow',
    });
  });

  wf.setHandler(gatePassedSignal, () => {
    gateAwaiting = false;
  });

  wf.setHandler(gateFailedSignal, async ({ reason }) => {
    isPaused = true;
    status = 'paused';
    gateAwaiting = false;
    await persistRunState({
      promotionRunId: input.promotionRunId,
      status: 'paused',
      pauseReason: reason,
    });
  });

  await persistRunState({
    promotionRunId: input.promotionRunId,
    status: 'active',
  });
  await recordAuditEvent({
    promotionRunId: input.promotionRunId,
    action: 'run_started',
    actorType: input.actor.actorType,
    actorId: input.actor.actorId,
    displayName: input.actor.displayName,
  });

  while (currentStageIndex < input.stageCount && status !== 'aborted') {
    await wf.condition(() => !isPaused || status === 'aborted');
    if (status === 'aborted') {
      break;
    }

    await recordAuditEvent({
      promotionRunId: input.promotionRunId,
      action: 'stage_entered',
      actorType: 'system',
      actorId: 'workflow',
      metadata: { stageIndex: currentStageIndex },
    });

    const gateResult = await evaluateGate({
      promotionRunId: input.promotionRunId,
      stageIndex: currentStageIndex,
    });

    if (gateResult.verdict === 'fail') {
      isPaused = true;
      status = 'paused';
      await persistRunState({
        promotionRunId: input.promotionRunId,
        status: 'paused',
      });
      await wf.condition(() => !isPaused || status === 'aborted');
      if (status === 'aborted') {
        break;
      }
      continue;
    }

    const nextStageIndex = currentStageIndex + 1;
    await persistRunState({
      promotionRunId: input.promotionRunId,
      status: 'active',
      currentStageIndex: nextStageIndex,
    });
    currentStageIndex = nextStageIndex;

    await recordAuditEvent({
      promotionRunId: input.promotionRunId,
      action: 'stage_advanced',
      actorType: 'system',
      actorId: 'workflow',
      metadata: { stageIndex: currentStageIndex },
      gateResultId: gateResult.gateResultId,
    });
  }

  if (status !== 'aborted') {
    status = 'completed';
    await persistRunState({
      promotionRunId: input.promotionRunId,
      status: 'completed',
    });
    await recordAuditEvent({
      promotionRunId: input.promotionRunId,
      action: 'run_completed',
      actorType: 'system',
      actorId: 'workflow',
    });
  }
}

File: kotlin/modules/worker/src/main/kotlin/com/ffpromo/worker/workflows/PromotionWorkflow.kt
package com.ffpromo.worker.workflows

import com.ffpromo.contracts.AuditAction
import com.ffpromo.contracts.ActorType
import com.ffpromo.contracts.PromotionStatus
import com.ffpromo.worker.activities.EvaluateGateActivity
import com.ffpromo.worker.activities.PersistRunStateActivity
import com.ffpromo.worker.activities.RecordAuditEventActivity
import io.temporal.activity.ActivityOptions
import io.temporal.common.RetryOptions
import io.temporal.workflow.SignalMethod
import io.temporal.workflow.Workflow
import io.temporal.workflow.WorkflowInterface
import io.temporal.workflow.WorkflowMethod
import java.time.Duration

@WorkflowInterface
interface PromotionWorkflow {
    @WorkflowMethod
    fun run(promotionRunId: String, pipelineId: String, flagKey: String)

    @SignalMethod(name = PromotionSignals.PAUSE)
    fun pause()

    @SignalMethod(name = PromotionSignals.RESUME)
    fun resume()

    @SignalMethod(name = PromotionSignals.ABORT)
    fun abort()
}

class PromotionWorkflowImpl : PromotionWorkflow {
    private val activityOptions = ActivityOptions.newBuilder()
        .setStartToCloseTimeout(Duration.ofSeconds(30))
        .setRetryOptions(RetryOptions.newBuilder().setMaximumAttempts(3).build())
        .build()

    private val persistRunState: PersistRunStateActivity =
        Workflow.newActivityStub(PersistRunStateActivity::class.java, activityOptions)
    private val recordAuditEvent: RecordAuditEventActivity =
        Workflow.newActivityStub(RecordAuditEventActivity::class.java, activityOptions)
    private val evaluateGate: EvaluateGateActivity =
        Workflow.newActivityStub(EvaluateGateActivity::class.java, activityOptions)

    private var currentStageIndex = 0
    private var status = "active"
    private var isPaused = false
    private var pauseRequested = false
    private var resumeRequested = false
    private var abortRequested = false
    private val stageCount = 3
    private var workflowPromotionRunId = ""

    override fun pause() {
        pauseRequested = true
    }

    override fun resume() {
        resumeRequested = true
    }

    override fun abort() {
        abortRequested = true
    }

    override fun run(promotionRunId: String, pipelineId: String, flagKey: String) {
        workflowPromotionRunId = promotionRunId

        persistRunState.persistRunState(
            promotionRunId,
            PromotionStatus.active.name,
            null,
            null,
        )
        recordAuditEvent.recordAuditEvent(
            promotionRunId,
            AuditAction.run_started.name,
            ActorType.system.name,
            "workflow",
            null,
            null,
            null,
        )

        while (currentStageIndex < stageCount && !hasAborted()) {
            applyPendingSignals()
            if (hasAborted()) break

            if (isPaused) {
                Workflow.await(Duration.ofDays(365)) { resumeRequested || abortRequested }
                applyPendingSignals()
                if (hasAborted()) break
                if (isPaused) continue
            }

            recordAuditEvent.recordAuditEvent(
                promotionRunId,
                AuditAction.stage_entered.name,
                ActorType.system.name,
                "workflow",
                null,
                null,
                """{"stageIndex":$currentStageIndex}""",
            )

            val gateResultId = evaluateGate.evaluateGate(promotionRunId, currentStageIndex)

            currentStageIndex += 1
            persistRunState.persistRunState(
                promotionRunId,
                PromotionStatus.active.name,
                currentStageIndex,
                null,
            )
            recordAuditEvent.recordAuditEvent(
                promotionRunId,
                AuditAction.stage_advanced.name,
                ActorType.system.name,
                "workflow",
                null,
                gateResultId,
                """{"stageIndex":$currentStageIndex}""",
            )
        }

        if (!hasAborted()) {
            status = "completed"
            persistRunState.persistRunState(
                promotionRunId,
                PromotionStatus.completed.name,
                null,
                null,
            )
            recordAuditEvent.recordAuditEvent(
                promotionRunId,
                AuditAction.run_completed.name,
                ActorType.system.name,
                "workflow",
                null,
                null,
                null,
            )
        }
    }

    private fun applyPendingSignals() {
        if (abortRequested) {
            abortRequested = false
            status = "aborted"
            persistRunState.persistRunState(
                workflowPromotionRunId,
                PromotionStatus.aborted.name,
                null,
                null,
            )
            recordAuditEvent.recordAuditEvent(
                workflowPromotionRunId,
                AuditAction.run_aborted.name,
                ActorType.system.name,
                "workflow",
                null,
                null,
                null,
            )
        }

        if (pauseRequested) {
            pauseRequested = false
            isPaused = true
            status = "paused"
            persistRunState.persistRunState(
                workflowPromotionRunId,
                PromotionStatus.paused.name,
                null,
                null,
            )
            recordAuditEvent.recordAuditEvent(
                workflowPromotionRunId,
                AuditAction.run_paused.name,
                ActorType.system.name,
                "workflow",
                null,
                null,
                null,
            )
        }

        if (resumeRequested) {
            resumeRequested = false
            isPaused = false
            status = "active"
            persistRunState.persistRunState(
                workflowPromotionRunId,
                PromotionStatus.active.name,
                null,
                null,
            )
            recordAuditEvent.recordAuditEvent(
                workflowPromotionRunId,
                AuditAction.run_resumed.name,
                ActorType.system.name,
                "workflow",
                null,
                null,
                null,
            )
        }
    }

    private fun hasAborted(): Boolean = status == "aborted"
}
